### ISSUE TITLE
Invalid output file

### ISSUE THREAD

**Issue by dcdamien** [Ideolys/carbone#1]
Hi,
I'm using following _data_: https://gist.github.com/dcdamien/9e50948219d3792307521c43b3108877

And following templates:

1.

```
{d.type}
{d.numberAssignments[i=0].designation}
{d.seller.address.phoneNumbers[i=0].number}
```

2.

| itemName |
| -- |
| {d.lines[i].itemName} |
| {d.lines[i+1].itemName} |

Carbone generates correct output files only when using single of them, but when they are merged into one template, output file gets corrupted. Opening file in Microsoft Word ends with `The file cannot be opened because there are problems with the contents` error.

[output-files.zip](https://github.com/Ideolys/carbone/files/1253253/output-files.zip)

**Comment by dcdamien** [comment]
I have done further tests and unfortunately looks like there is more issues.
Using following template:
```
{d.buyer.name}
{d.numberAssignments[i=1].designation}
```

| **approveDate** |
| --- |
|  {d.approveDate:convDate("YYYYMMDD", "LL")}  |

| **itemName** |
| --- |
| {d.lines[i].itemName} |
| {d.lines[i+1].itemName} |


Result looks as below:

![image](https://user-images.githubusercontent.com/7643344/29770322-cd1940a2-8bee-11e7-9915-f54538ec9a77.png)

Original files attached: [original-files.zip](https://github.com/Ideolys/carbone/files/1256527/original-files.zip)

**Comment by dgrelaud** [comment]
Hi, thank you for your samples. I can reproduce it. 🤔  Strange... quite simple case.

I will fix it as soon as I can this week...

**Comment by dgrelaud** [comment]
In fact, I know this bug!

@dcdamien  there is a workaround:
Create a "horizontal repetition block" `[i] ...[i+1]` and use the filter `i=0` in the same time for each array, like this:

```
{d.type}
{d.numberAssignments[i, i=0].designation}{d.numberAssignments[i+1, i=0].designation}
{d.seller.address.phoneNumbers[i, i=0].number}{d.seller.address.phoneNumbers[i+1, i=0].number}
```

The solution is not intuitive but the result is the same.
Is it ok for you? 

When we use a filter `[i=0]` (without repetition), it affects other repetitions... I know where the problem is but the solution is not simple. Again, I hope I will fix it this week to close definitely this issue

**Comment by dcdamien** [comment]
@dgrelaud i'm not sure if it workaround is completely effective.
Please look at following attempts:

**1.**
template:

```
{d.buyer.name}
{d.numberAssignments[i, i=0].designation} {d.numberAssignments[i+1, i=0].designation}
{d.seller.address.phoneNumbers[i, i=0].number} {d.seller.address.phoneNumbers[i+1, i=0].number}
```
| **approveDate** |
| --- |
|  {d.approveDate:convDate("YYYYMMDD", "LL")}  |

| **itemName** |
| --- |
| {d.lines[i].itemName} |
| {d.lines[i+1].itemName} |

output:

![image](https://user-images.githubusercontent.com/7643344/29776412-d3405622-8c08-11e7-963c-3334fdf8fa3c.png)


**2.**
template:
```
{d.buyer.name}
{d.numberAssignments[i, i=0].designation} {d.numberAssignments[i+1, i=0].designation}
```
| **approveDate** |
| --- |
|  {d.approveDate:convDate("YYYYMMDD", "LL")}  |

| **itemName** |
| --- |
| {d.lines[i].itemName} |
| {d.lines[i+1].itemName} |

output: [**corrupted-file**](https://github.com/Ideolys/carbone/files/1256933/corrupted-file.zip)

Please note that only difference is occurrence of 3rd line: `{d.seller.address.phoneNumbers[i, i=0].number} {d.seller.address.phoneNumbers[i+1, i=0].number}`

I'll patiently wait for bugfix.

**Comment by dgrelaud** [comment]
The last week was not long enough to fix this bug... I do not forget you ;)

**Comment by dcdamien** [comment]
Hi @dgrelaud. Hope you're doing well. Have you any schedule on next release which will include fix?

**Comment by dgrelaud** [comment]
Hi,

Sorry, I'm very busy currently. I was a bit optimist... I hope to fix it at the beginning of October.

**Comment by dcdamien** [comment]
@dgrelaud if you know this bug maybe can you give some more information about it? We can try to fix it ourselves..

**Comment by dcdamien** [comment]
Hi @dgrelaud, any news? hints?

**Comment by dgrelaud** [comment]
Hi,

>  if you know this bug maybe can you give some more information about it? We can try to fix it ourselves..

Yes, good idea. But it is hard to explain 😞 .

The philosophy of Carbone is to generate JS code for each report (and cache it => todo) in order to build the report as fast as possible. This part is not easy.

Roughly, when carbone sees  `d.myArray[i=1]`, it considers there is an XML row to repeat whereas it is not the case. It should generate JS code to iterate `myArray`, filter and find the row which matches with the filter `i=1` and generate XML as if it was an object.

You can try to add your test case in `test/test.builder.buildXML.js` 

The problem is in `lib/builder.js` in `getBuilderFunction`.

Sorry, I was a bit optimist on my agenda.

**Comment by normowl** [comment]
Is there a workaround for array based data?

**Comment by dgrelaud** [comment]
Just some news. I am not dead, just very busy.

This bug is solved internally!
I have to clean the code, push ans publish everything. 

Stay tuned.

**Comment by dgrelaud** [comment]
fixed in v1.1.0
